### DISCUSSION TITLE
Requiring a Type Class in Enum Definition

### DISCUSSION THREAD

**Discussion by ashwinbhaskar** [scala/scala3#12639] [General Discussion]
For the code below, the compiler complains `[error]   |No implicit Ordering defined for A..`
```
import scala.math.Ordering

enum BinaryTree[+A: Ordering]:
    case Leaf
    case Node(left: BinaryTree[A], value: A, right: BinaryTree[A])

```
The reason for this error was pointed out [in this post](https://users.scala-lang.org/t/encoding-a-binary-tree-in-scala-3/7474). 

But for someone who doesn't know what enums desugar to, the error might be puzzling because they have already said `A has Ordering` in the enum declaration.

is this a valid complaint?

**Comment by bishabosha** [comment]
i think it would be useful to also copy over any context bounds in the desugaring, which would have to be a change in a future minor version.

It is worth noting though that by adding an `Ordering` context bound - your `BinaryTree` now has hidden state that makes it harder to program generically. e.g. there will be no `scala.deriving.Mirror` generated

**Comment by nicolasstucki** [comment]
__Workarround__
```diff
import scala.math.Ordering
enum BinaryTree[+A: Ordering]:
    case Leaf
-    case Node(left: BinaryTree[A], value: A, right: BinaryTree[A])
+    case Node(left: BinaryTree[A], value: A, right: BinaryTree[A])(using Ordering[A])
```
